### ISSUE TITLE
NullPointerException in ArcContextProvider#deactivateRequestContext on application shutdown

### ISSUE THREAD

**Issue by yuriytkach** [quarkusio/quarkus#14980]
**Describe the bug**
We have a simple QuarkusApplication with QuarkusMain started as a cronjob 4 times a day. Application uses the following components: quarkus-rest-client, quarkus-vault, quarkus-jackson, quarkus-smallrye-reactive-messaging-kafka, quarkus-vertx.
It is running in JVM mode.
It simply downloads some data from rest api, produces kafka message and shuts down.
Sometimes, after the last application log message is written (ie, all tasks were finished), we see the following error message in logs at the end:
```
message: Mutiny had to drop the following exception
exception.exceptionType | java.lang.NullPointerException
```
And the following exception frames:
```
{
  "class": "io.quarkus.arc.runtime.context.ArcContextProvider",
  "method": "deactivateRequestContext",
  "line": 72
},
{
  "class": "io.quarkus.arc.runtime.context.ArcContextProvider",
  "method": "access$200",
  "line": 20
},
{
  "class": "io.quarkus.arc.runtime.context.ArcContextProvider$ContextSnapshot",
  "method": "lambda$begin$0",
  "line": 158
},
{
  "class": "io.smallrye.context.impl.ActiveContextState",
  "method": "close",
  "line": 28
},
{
  "class": "io.smallrye.context.SmallRyeThreadContext",
  "method": "lambda$withContext$1",
  "line": 531
},
{
  "class": "io.smallrye.mutiny.context.ContextPropagationUniInterceptor$1",
  "method": "onItem",
  "line": 31
},
{
  "class": "io.smallrye.mutiny.operators.UniSerializedSubscriber",
  "method": "onItem",
  "line": 86
},
{
  "class": "io.smallrye.mutiny.vertx.AsyncResultUni",
  "method": "lambda$subscribing$1",
  "line": 34
},
{
  "class": "io.vertx.core.impl.ContextImpl",
  "method": "lambda$null$0",
  "line": 327
},
{
  "class": "io.vertx.core.impl.ContextImpl",
  "method": "executeTask",
  "line": 366
},
{
  "class": "io.vertx.core.impl.EventLoopContext",
  "method": "lambda$executeAsync$0",
  "line": 38
},
{
  "class": "io.netty.util.concurrent.AbstractEventExecutor",
  "method": "safeExecute",
  "line": 164
},
{
  "class": "io.netty.util.concurrent.SingleThreadEventExecutor",
  "method": "runAllTasks",
  "line": 472
},
{
  "class": "io.netty.channel.nio.NioEventLoop",
  "method": "run",
  "line": 500
},
{
  "class": "io.netty.util.concurrent.SingleThreadEventExecutor$4",
  "method": "run",
  "line": 989
},
{
  "class": "io.netty.util.internal.ThreadExecutorMap$2",
  "method": "run",
  "line": 74
},
{
  "class": "io.netty.util.concurrent.FastThreadLocalRunnable",
  "method": "run",
  "line": 30
},
{
  "class": "java.lang.Thread",
  "method": "run",
  "line": 834
}
```

Application is run as

**Environment (please complete the following information):**
 - Producing docker image with `Dockerfile.jvm` that came from start.quarkus.io without any modification (FROM registry.access.redhat.com/ubi8/ubi-minimal:8.3, ARG JAVA_PACKAGE=java-11-openjdk-headless, ARG RUN_JAVA_VERSION=1.3.8)
 - Quarkus version or git rev: *1.11.1.Final*
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 3.6.3

**Comment by stuartwdouglas** [comment]
How are you shutting down Quarkus? It sounds like a request is still in progress when shutdown is triggered.

**Comment by yuriytkach** [comment]
> How are you shutting down Quarkus? It sounds like a request is still in progress when shutdown is triggered.

Nothing special, I just return from the main method:
```
@Override
public int run(final String... args) {
   ...
   return 0;
}
```

Are there any specific shutdown methods that I should call for proper shutdown?

Interesting, that out of 4 runs of the app during a day, only 1 run actually fails with NPE and not every day. I'm still puzzled to understand what conditions could lead to that.

**Comment by stuartwdouglas** [comment]
Do you have any guard that the work you are doing in an async manner is done before you start shutdown?

I think the issue here is likely that you are done with your work, and then you start shutdown, but Quarkus is still running its cleanup code (namely deactivating the request scope). For HTTP based apps we track current requests and can pause them till they are done, but it is harder to do with custom work. 

This particular exception is harmless, however if there was transaction propagation involved its possible that you could have a similar issue with TX commit, which would not be harmless.

**Comment by stuartwdouglas** [comment]
For now the simplest workaround is likely to add Thread.sleep(100) just before you exit, to give Quarkus time to finish.

**Comment by yuriytkach** [comment]
> For now the simplest workaround is likely to add Thread.sleep(100) just before you exit, to give Quarkus time to finish.

Thank you for your reply. I will try with thread sleep and see how it goes.
The last action that we do is sending the Kafka message with Emmitter. Maybe Kafka shutdown is not going properly.

**Comment by cescoffier** [comment]
Do you have a reproducer? 
Kafka shutdown is async (as polling may be in progress), we may start closing everything but `poll` returns and continue a bit of work instead of stopping.

**Comment by stuartwdouglas** [comment]
https://github.com/quarkusio/quarkus/pull/15067 should help with the NPE

**Comment by yuriytkach** [comment]
> Do you have a reproducer?
> Kafka shutdown is async (as polling may be in progress), we may start closing everything but `poll` returns and continue a bit of work instead of stopping.

Sorry, unfortunately, no reproducer, as it happens randomly on the exact same code that is run with the exact same params several times a day.

**Comment by gsmet** [comment]
@yuriytkach did you have the chance to try 1.12.0.Final?

**Comment by yuriytkach** [comment]
> @yuriytkach did you have the chance to try 1.12.0.Final?

Not yet. For now I can't compile with 1.12.x because of tests issue: TestResource with WireMock server in it won't update the url in configs (trying to figure out what's going on, because the exact same test works perfectly on 1.11.6).

Actually, I've been running application on 1.11.6 for the last week and application runs 6 times a day now - never saw an error again. If error never appears in the next week or two, I guess we could close the issue :)

**Comment by yuriytkach** [comment]
I've updated to latest Quarkus 1.13.1.
With it I've received a different error on application close related to Kafka producer: `Producer closed while send in progress`.
That one is much more clear than NPE that was before :)
I was doing `emitter.send()` but didn't wait for completion and just exited from the app. I've added `.toCompletableFuture().join()` after `send()` and that resolved the issue.
So, I think I will close the issue
